task01/src/Vector.cpp

```diff
@@ -0,0 +1,318 @@
+	if (m_size != 0) {
+		return &m_array[0];
+	}
+}
+
+//--------------------------------------------------------------------------------
+void Vector::clear() {
+	m_size = 0;
+	m_array = 0;
+}
+
+//--------------------------------------------------------------------------------
+
+void Vector::pop_back() {
+    if (m_size > 0) {
+        --m_size;
+    }
+}
+
+//--------------------------------------------------------------------------------
+/*check if left = right */
+bool operator==(const Vector& lhs, const Vector& rhs) {
+    if (lhs.getSize() != rhs.getSize()) {
+        return false;
+    }
+
+    for (std::size_t i = 0; i < lhs.getSize(); ++i) {
+        if (lhs[i] != rhs[i]) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+//--------------------------------------------------------------------------------
+/*this function inserts a value in a giving possision.*/
+int* Vector::insert(iterator pos, const int& val) {
+    // calculate the index of the position iterator
+    int index = pos - begin();
+
+    // if the new size exceeds the current capacity, resize the array
+    if (m_size + 1 > m_capacity) {
+        m_capacity = (m_capacity == 0 ? 1 : m_capacity * 2);
+        int* newArray = new int[m_capacity];
+        // copy the elements before the insertion position
+        for (int i = 0; i < index; i++) {
+            newArray[i] = m_array[i];
+        }
+        // insert the new element
+        newArray[index] = val;
+        // copy the elements after the insertion position
+        for (int i = index + 1; i < m_size + 1; i++) {
+            newArray[i] = m_array[i - 1];
+        }
+        // update the size and delete the old array
+        m_size++;
+        delete[] m_array;
+        m_array = newArray;
+    }
+    // if the new size doesn't exceed the current capacity, insert the element in-place
+    else {
+        // move the elements after the insertion position to make space for the new element
+        for (int i = m_size; i > index; i--) {
+            m_array[i] = m_array[i - 1];
+        }
+        // insert the new element
+        m_array[index] = val;
+        // update the size
+        m_size++;
+    }
+
+    return pos;
+}
+
+//--------------------------------------------------------------------------------
+/*this functions swaps between two vectors.*/
+void Vector::swap(Vector& other) {
+    // Swap size
+    const std::size_t tempSize = m_size;
+    m_size = other.m_size;
+    other.m_size = tempSize;
+
+    // Swap capacity
+    const std::size_t tempCapacity = m_capacity;
+    m_capacity = other.m_capacity;
+    other.m_capacity = tempCapacity;
+
+    // Swap arrays
+    int* const tempArray = m_array;
+    m_array = other.m_array;
+    other.m_array = tempArray;
+}
+
+
+//--------------------------------------------------------------------------------
+/*tDelete element from an iterator*/
+int* Vector::erase(iterator pos) {
+    int* newArray = new int[m_size - 1];
```

Comment:
Why re-allocate the entire array? It is enough the delete the element from the existing array (and then just moving all the elements to the right)